### DISCUSSION TITLE
useProxy and useLayoutEffect

### DISCUSSION THREAD

**Discussion by juliancwirko** [pmndrs/valtio#652] [General]
I'm getting a warning regarding useEffectLayout and SSR on the server side in Next.js (the newest version). Are there any drawbacks to something like this? 

```jsx
const useIsomorphicEffect = typeof window !== 'undefined' ? useLayoutEffect : useEffect

export function useProxy<T extends object>(proxy: T): T {
  const snapshot = useSnapshot(proxy) as T

  let isRendering = true

  useIsomorphicEffect(() => {
    // eslint-disable-next-line react-hooks/exhaustive-deps
    isRendering = false
  })

  return new Proxy(proxy, {
    get(target, prop) {
      return isRendering ? snapshot[prop as keyof T] : target[prop as keyof T]
    },
  })
}
```

**Comment by dai-shi** [comment]
useEffect is no-op on server. And, it's very hard to detect if we are on sever.
If you can't live with the false positive warning, you could do something like this.

```js
import { snapshot } from 'valtio'
import { useProxy as useProxyOrig } from 'valtio/utils'

const isSSR = typeof window === 'undefined' // this is usually not enough
export const useProxy = isSSR ? (p) => snapshot(p) : useProxyOrig
```

**Reply by juliancwirko** [reply]
makes sense, thanks for the guidance

**Reply by SkylarWolfLA** [reply]
Hi I just wanted to re-open this discussion as we are using valtio 2.0.0 and loving it, but the warnings on the server-side with next.js are flooding our logs in development.

I saw that in use-context-selector there was a fix, and we could perhaps bring that fix here?

https://github.com/dai-shi/use-context-selector/blob/main/src/index.ts#L26

```
const isSSR =
  typeof window === 'undefined' ||
  /ServerSideRendering/.test(window.navigator && window.navigator.userAgent);

const useIsomorphicLayoutEffect = isSSR ? useEffect : useLayoutEffect;
```

**Reply by dai-shi** [reply]
https://github.com/pmndrs/valtio/discussions/890
https://github.com/pmndrs/valtio/pull/891

They are related discussions, and now remember why I was waiting for React 19 before releasing Valtio v2. So, we knew this and this is intentional change.

Is it correct that if you use newer Next.js, the warning won't be shown?

**Reply by dai-shi** [reply]
Does this work?

```js
import { snapshot, useSnapshot as useSnapshotOrig } from 'valtio'

const isSSR = typeof window === 'undefined' // this is usually not enough
export const useSnapshot = isSSR ? (p) => snapshot(p) : useSnapshotOrig
```

**Reply by SkylarWolfLA** [reply]
We are on Next 14.2.5 and having the warnings in dev mode so not sure if 14.2.7 or 15 has a fix for it but at least 14.2.5 doesn't.

I remember seeing that solution somewhere else but I didn't put it together for some reason that that would solve the useLayoutEffect warnings.

It does work perfectly though and I imagine it would be more performant as well so win-win!

Thank you for the fast response!

**Reply by aretrace** [reply]
> it's very hard to detect if we are on sever

would somthing [like](https://github.com/atlassian/react-beautiful-dnd/blob/ae7bcf326928fea3615dfed092dd4261a3976e0d/src/view/use-isomorphic-layout-effect.js#L11C1-L16C17):
```js
const useIsomorphicLayoutEffect =
  typeof window !== 'undefined' &&
  typeof window.document !== 'undefined' &&
  typeof window.document.createElement !== 'undefined'
    ? useLayoutEffect
    : useEffect;
```
be enough to introduce `useIsomorphicLayoutEffect` [into](https://github.com/pmndrs/valtio/blob/7734e7b5bf22c0b4db5310a3092d32b8b6dd3023/src/react/utils/useProxy.ts#L38) [valtio](https://github.com/pmndrs/valtio/blob/7734e7b5bf22c0b4db5310a3092d32b8b6dd3023/src/react.ts#L155) itself? 
using [valtio with astro](https://stackblitz.com/edit/withastro-astro-yofz6y?file=src%2Fcomponents%2Fcounter.tsx) and facebook/react#14363 is flooding the dev logs as well

**Reply by dai-shi** [reply]
Well, I was so happy to say good bye to `useIsomorphicLayoutEffect`: https://x.com/dai_shi/status/1636022438696980480
It was a wrong conception and I think `useIsomorphicLayoutEffect` misguided the community.

Does this work for you? `export const useSnapshot = isSSR ? (p) => snapshot(p) : useSnapshotOrig`

We can provide it as a special util, but I think it's better not to hide this problem, because it was a false positive and it will be fixed soon.

**Reply by aretrace** [reply]
`isSSR` [works](https://stackblitz.com/edit/withastro-astro-yofz6y?file=src%2Fcomponents%2Fcounter.tsx) 👍
ty for pointing out  facebook/react#26395, I'll use the workaround for now but I imagine others stuck on older versions in the future would appreciate a util for this.

**Reply by dai-shi** [reply]
Let's add it in the migration guide for now.
Would you like to send a PR?

**Reply by aretrace** [reply]
👉 pmndrs/valtio#944
